File: LaoCai_2324/Bai2/bai2.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <algorithm>

using namespace std;

// Hàm tính tích các thừa số nguyên tố khác nhau của n
long long calculateS(int n, const vector<int>& primes) {
    long long s = 1;
    for (int prime : primes) {
        if (prime > n) break; // Dừng nếu số nguyên tố lớn hơn n
        if (n % prime == 0) { // Nếu prime là thừa số của n
            s *= prime; // Nhân vào tích S
            // Loại bỏ tất cả các lần xuất hiện của prime trong n
            while (n % prime == 0) {
                n /= prime;
            }
        }
    }
    return s;
}

// Hàm chính
int main() {
    ifstream infile("bai2.inp");
    ofstream outfile("bai2.out");

    int M;
    infile >> M;

    // Sàng Eratosthenes để tìm tất cả các số nguyên tố <= M
    vector<bool> is_prime(M + 1, true);
    vector<int> primes;

    is_prime[0] = is_prime[1] = false; // 0 và 1 không phải là số nguyên tố
    for (int i = 2; i <= M; ++i) {
        if (is_prime[i]) {
            primes.push_back(i);
            for (int j = 2 * i; j <= M; j += i) {
                is_prime[j] = false;
            }
        }
    }

    // Tìm số N ≤ M có S lớn nhất
    int resultN = 0;
    long long maxS = 0;

    for (int i = 2; i <= M; ++i) {
        long long s = calculateS(i, primes);
        // Cập nhật nếu tìm thấy giá trị S lớn hơn
        if (s > maxS) {
            maxS = s;
            resultN = i;
        } else if (s == maxS && i > resultN) {
            resultN = i; // Lưu số lớn nhất nếu S bằng nhau
        }
    }

    // Ghi kết quả vào tệp
    outfile << resultN << endl;

    infile.close();
    outfile.close();
    return 0;
}

File: LaoCai_2324/Bai2/bai2.py
def sieve_of_eratosthenes(limit):
    """Hàm sử dụng Sàng Eratosthenes để tìm tất cả các số nguyên tố <= limit."""
    is_prime = [True] * (limit + 1)
    primes = []
    is_prime[0], is_prime[1] = False, False  # 0 và 1 không phải là số nguyên tố

    for i in range(2, limit + 1):
        if is_prime[i]:
            primes.append(i)
            for j in range(2 * i, limit + 1, i):
                is_prime[j] = False

    return primes


def calculate_s(n, primes):
    """Hàm tính tích các thừa số nguyên tố khác nhau của n."""
    s = 1
    for prime in primes:
        if prime > n:
            break
        if n % prime == 0:
            s *= prime
            # Loại bỏ tất cả các lần xuất hiện của prime trong n
            while n % prime == 0:
                n //= prime
    return s


def main():
    # Đọc giá trị M từ tệp bai2.inp
    with open("bai2.inp", "r") as infile:
        M = int(infile.readline().strip())

    # Tìm tất cả các số nguyên tố <= M
    primes = sieve_of_eratosthenes(M)

    # Tìm số N ≤ M có S lớn nhất
    result_n = 0
    max_s = 0

    for i in range(2, M + 1):
        s = calculate_s(i, primes)
        # Cập nhật nếu tìm thấy giá trị S lớn hơn
        if s > max_s:
            max_s = s
            result_n = i
        elif s == max_s and i > result_n:
            result_n = i  # Lưu số lớn nhất nếu S bằng nhau

    # Ghi kết quả vào tệp bai2.out
    with open("bai2.out", "w") as outfile:
        outfile.write(f"{result_n}\n")


if __name__ == "__main__":
    main()
